libraries/DateTime/contracts/DateTime.aes

```diff
@@ -0,0 +1,244 @@
+contract DateTime =
+  /*
+   Date and Time utilities for sophia contracts
+  */
+
+// Define used structures
+  record state = 
+    {
+      origin_year : int, 
+      day_in_seconds : int,
+      year_in_seconds : int,
+      leap_year_in_seconds : int,
+      hour_in_seconds : int,
+      minute_in_seconds : int
+    }
+
+  record date_time = 
+    {
+      year : int,
+      month : int,
+      day : int,
+      hour : int,
+      minute : int,
+      second : int,
+      weekday : int,
+      timestamp : int
+    }
+  
+  record buff = 
+    {
+      year : int,
+      month : int,
+      day : int,
+      leap_years_before : int,
+      seconds_accounted_for : int,
+      timestamp : int
+    }                    
+
+// Public function (API)
+  public function init() : state =
+    {
+      origin_year = 1970, 
+      day_in_seconds = 86400,
+      year_in_seconds = 31536000,
+      leap_year_in_seconds = 31622400,
+      hour_in_seconds = 3600,
+      minute_in_seconds = 60
+      }
+
+  public function parse_timestamp(timestamp : int) : date_time = 
+    let dt : date_time = init_date_time(timestamp)
+    let buf : buff = init_buff(timestamp)
+
+    // calculate year by given timestamp
+    let buf : buff = p_get_year(buf)
+    let dt : date_time = dt{year = buf.year}
+
+    // calculate month by given timestamp
+    let buf : buff = p_get_month(buf)
+    let dt : date_time = dt{month = buf.month}
+
+    // calculate day by given timestamp
+    let buf : buff = p_get_day(buf)
+    let dt : date_time = dt{day = buf.day}
+
+    // calculate hour by given timestamp
+    let dt : date_time = dt{hour = rest(timestamp / 60 /60, 24)}
+
+    // calculate minute by given timestamp
+    let dt : date_time = dt{minute = rest(timestamp / 60, 60)}
+
+    // calculate second by given timestamp
+    let dt : date_time = dt{second = rest(timestamp, 60)}
+
+    // calculate weekday by given timestamp
+    let wd : int = rest(timestamp / state.day_in_seconds + 4, 7)
+    let dt : date_time = dt{weekday = wd}
+    dt
+
+  public function get_year(timestamp : int) : int =
+    parse_timestamp(timestamp).year
```

Comment:
maybe you should separate all these calculations. if you get year(get_year) compute only year... if you get try to get month calculate year and month ... and so on

Reply:
Yes, you are right!